src/osd/PG.h

```diff
@@ -77,8 +77,8 @@ namespace Scrub {
   class Store;
 }
 
-using state_history_entry = std::tuple<utime_t, utime_t, const char*>;
-using embedded_state = std::pair<utime_t, const char*>;
+using state_history_entry = std::tuple<utime_t, utime_t, const std::string>;
+using embedded_state = std::pair<utime_t, const std::string>;
```

Comment:
Ah, sorry I forgot that this worked that way.  The const char* here is simply a pointer pointing into the data section of the process.  By switching it to std::string, you are causing a string copy and allocation to happen from that data section into the heap.  Instead, you probably just want to make get_current_state return a const char*.

Reply:
Yep, didn't think it was an efficiency win but got sucked down the rat hole.